File: src-tauri/src/twitch/moderation.rs
use super::{get_token, helix};
use serde::Deserialize;
use std::borrow::Cow;
use twitch_api::helix::channels::{AddChannelVipRequest, RemoveChannelVipRequest};
use twitch_api::helix::moderation::{
    delete_chat_messages::DeleteChatMessagesRequest,
    manage_held_automod_messages::{
        ManageHeldAutoModMessagesBody, ManageHeldAutoModMessagesRequest,
    },
    warn_chat_user::{WarnChatUserBody, WarnChatUserRequest},
    BanUser, BanUserBody, BanUserRequest, BannedUser, GetBannedUsersRequest,
    GetModeratedChannelsRequest, GetModeratorsRequest, ModeratedChannel, Moderator,
    UnbanUserRequest,
};
use twitch_api::helix::raids::{CancelARaidRequest, StartARaidRequest};
use twitch_api::helix::{Cursor, EmptyBody};
use twitch_api::types::MsgId;

use super::response::PaginatedResponse;

// https://dev.twitch.tv/docs/api/reference/#delete-chat-messages
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteChatMessagesParams {
    pub broadcaster_id: String,
    pub message_id: Option<String>,
}

#[tauri::command]
pub async fn delete_chat_messages(
    app: tauri::AppHandle,
    params: DeleteChatMessagesParams,
) -> Result<(), String> {
    let token = get_token(&app).await?;

    let mut request = DeleteChatMessagesRequest::new(
        params.broadcaster_id.as_str(),
        token.user_id.as_str(),
    );
    request.message_id = params.message_id.map(|s| Cow::Owned(MsgId::from(s)));

    helix()
        .req_delete(request, &token)
        .await
        .map(|_| ())
        .map_err(|e| e.to_string())
}
// https://dev.twitch.tv/docs/api/reference/#ban-user
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BanUserParams {
    pub broadcaster_id: String,
    pub user_id: String,
    #[serde(default)]
    pub duration: Option<u32>,
    #[serde(default)]
    pub reason: Option<String>,
}

#[tauri::command]
pub async fn ban_user(app: tauri::AppHandle, params: BanUserParams) -> Result<BanUser, String> {
    let token = get_token(&app).await?;
    let request = BanUserRequest::new(params.broadcaster_id.as_str(), token.user_id.as_str());
    let body = BanUserBody::new(
        params.user_id.as_str(),
        params.reason.unwrap_or_default(),
        params.duration,
    );
    helix()
        .req_post(request, body, &token)
        .await
        .map(|r| r.data)
        .map_err(|e| e.to_string())
}

// https://dev.twitch.tv/docs/api/reference/#get-banned-users
#[derive(Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct GetBannedUsersParams {
    pub broadcaster_id: String,
    pub user_id: Option<String>,
    pub first: Option<usize>,
    pub after: Option<String>,
}

#[tauri::command]
pub async fn get_banned_users(
    app: tauri::AppHandle,
    params: GetBannedUsersParams,
) -> Result<PaginatedResponse<BannedUser>, String> {
    let token = get_token(&app).await?;
    let mut request = GetBannedUsersRequest::broadcaster_id(params.broadcaster_id.as_str());
    if let Some(uid) = params.user_id.as_deref() {
        request.user_id = vec![twitch_api::types::UserId::from(uid)].into();
    }
    request.first = params.first;
    request.after = params.after.map(|s| Cow::Owned(Cursor::from(s)));

    let response = helix()
        .req_get(request, &token)
        .await
        .map_err(|e| e.to_string())?;
    Ok(PaginatedResponse::new(response.data, response.pagination))
}

#[derive(Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct GetAllBannedUsersParams {
    pub broadcaster_id: String,
    pub user_id: Option<String>,
}

#[tauri::command]
pub async fn get_all_banned_users(
    app: tauri::AppHandle,
    params: GetAllBannedUsersParams,
) -> Result<Vec<BannedUser>, String> {
    let token = get_token(&app).await?;
    let bc = params.broadcaster_id;
    let uid = params.user_id;
    super::utils::fetch_all_pages(&token, |after| {
        let mut request =
            GetBannedUsersRequest::broadcaster_id(twitch_api::types::UserId::from(bc.as_str()));
        if let Some(u) = uid.as_deref() {
            request.user_id = vec![twitch_api::types::UserId::from(u)].into();
        }
        request.after = after.map(|s| Cow::Owned(Cursor::from(s)));
        request
    })
    .await
}

// https://dev.twitch.tv/docs/api/reference/#unban-user
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UnbanUserParams {
    pub broadcaster_id: String,
    pub user_id: String,
}

#[tauri::command]
pub async fn unban_user(app: tauri::AppHandle, params: UnbanUserParams) -> Result<(), String> {
    let token = get_token(&app).await?;
    let request = UnbanUserRequest::new(
        params.broadcaster_id.as_str(),
        token.user_id.as_str(),
        params.user_id.as_str(),
    );
    helix()
        .req_delete(request, &token)
        .await
        .map(|_| ())
        .map_err(|e| e.to_string())
}

// https://dev.twitch.tv/docs/api/reference/#get-moderators
#[derive(Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct GetModeratorsParams {
    pub broadcaster_id: String,
    pub first: Option<usize>,
    pub after: Option<String>,
}

#[tauri::command]
pub async fn get_moderators(
    app: tauri::AppHandle,
    params: GetModeratorsParams,
) -> Result<PaginatedResponse<Moderator>, String> {
    let token = get_token(&app).await?;

    let mut request = GetModeratorsRequest::broadcaster_id(params.broadcaster_id.as_str());
    request.first = params.first;
    request.after = params.after.map(|s| Cow::Owned(Cursor::from(s)));

    let response = helix()
        .req_get(request, &token)
        .await
        .map_err(|e| e.to_string())?;

    Ok(PaginatedResponse::new(response.data, response.pagination))
}

#[derive(Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct GetAllModeratorsParams {
    pub broadcaster_id: String,
}

#[tauri::command]
pub async fn get_all_moderators(
    app: tauri::AppHandle,
    params: GetAllModeratorsParams,
) -> Result<Vec<Moderator>, String> {
    let token = get_token(&app).await?;
    let bc = params.broadcaster_id;
    super::utils::fetch_all_pages(&token, |after| {
        let mut request =
            GetModeratorsRequest::broadcaster_id(twitch_api::types::UserId::from(bc.as_str()));
        request.after = after.map(|s| Cow::Owned(Cursor::from(s)));
        request
    })
    .await
}

// https://dev.twitch.tv/docs/api/reference/#add-channel-vip
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AddChannelVipParams {
    pub broadcaster_id: String,
    pub user_id: String,
}

#[tauri::command]
pub async fn add_channel_vip(
    app: tauri::AppHandle,
    params: AddChannelVipParams,
) -> Result<(), String> {
    let token = get_token(&app).await?;
    let request =
        AddChannelVipRequest::new(params.broadcaster_id.as_str(), params.user_id.as_str());
    helix()
        .req_post(request, twitch_api::helix::EmptyBody, &token)
        .await
        .map(|_| ())
        .map_err(|e| e.to_string())
}

// https://dev.twitch.tv/docs/api/reference/#remove-channel-vip
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RemoveChannelVipParams {
    pub broadcaster_id: String,
    pub user_id: String,
}

#[tauri::command]
pub async fn remove_channel_vip(
    app: tauri::AppHandle,
    params: RemoveChannelVipParams,
) -> Result<(), String> {
    let token = get_token(&app).await?;
    let request =
        RemoveChannelVipRequest::new(params.broadcaster_id.as_str(), params.user_id.as_str());
    helix()
        .req_delete(request, &token)
        .await
        .map(|_| ())
        .map_err(|e| e.to_string())
}

// https://dev.twitch.tv/docs/api/reference/#get-moderated-channels
#[derive(Default, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct GetModeratedChannelsParams {
    pub first: Option<usize>,
    pub after: Option<String>,
}

#[tauri::command]
pub async fn get_moderated_channels(
    app: tauri::AppHandle,
    params: GetModeratedChannelsParams,
) -> Result<PaginatedResponse<ModeratedChannel>, String> {
    let token = get_token(&app).await?;

    let mut request = GetModeratedChannelsRequest::user_id(token.user_id.clone());
    request.first = params.first;
    request.after = params.after.map(|s| Cow::Owned(Cursor::from(s)));

    let response = helix()
        .req_get(request, &token)
        .await
        .map_err(|e| e.to_string())?;

    Ok(PaginatedResponse::new(response.data, response.pagination))
}

/// Fetches every page of moderated channels for the authenticated user and,
/// as a side effect, refreshes the `moderated_channel_ids` cache used by the
/// eventsub task to decide `is_mod` for new subscriptions.
#[tauri::command]
pub async fn get_all_moderated_channels(
    app: tauri::AppHandle,
) -> Result<Vec<ModeratedChannel>, String> {
    let token = get_token(&app).await?;
    let channels = super::utils::fetch_all_pages(&token, |after| {
        let mut request = GetModeratedChannelsRequest::user_id(token.user_id.clone());
        request.after = after.map(|s| Cow::Owned(Cursor::from(s)));
        request
    })
    .await?;
    super::utils::cache_moderated_channel_ids(&app, &channels);
    Ok(channels)
}

// https://dev.twitch.tv/docs/api/reference/#start-a-raid
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StartRaidParams {
    pub from_broadcaster_id: String,
    pub to_broadcaster_id: String,
}

#[tauri::command]
pub async fn start_raid(
    app: tauri::AppHandle,
    params: StartRaidParams,
) -> Result<(), String> {
    let token = get_token(&app).await?;
    let request = StartARaidRequest::new(
        params.from_broadcaster_id.as_str(),
        params.to_broadcaster_id.as_str(),
    );
    helix()
        .req_post(request, EmptyBody, &token)
        .await
        .map(|_| ())
        .map_err(|e| e.to_string())
}

// https://dev.twitch.tv/docs/api/reference/#cancel-a-raid
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CancelRaidParams {
    pub broadcaster_id: String,
}

#[tauri::command]
pub async fn cancel_raid(
    app: tauri::AppHandle,
    params: CancelRaidParams,
) -> Result<(), String> {
    let token = get_token(&app).await?;
    let request = CancelARaidRequest::broadcaster_id(params.broadcaster_id.as_str());
    helix()
        .req_delete(request, &token)
        .await
        .map(|_| ())
        .map_err(|e| e.to_string())
}

// https://dev.twitch.tv/docs/api/reference/#warn-chat-user
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WarnUserParams {
    pub broadcaster_id: String,
    pub user_id: String,
    pub reason: String,
}

#[tauri::command]
pub async fn warn_user(app: tauri::AppHandle, params: WarnUserParams) -> Result<(), String> {
    let token = get_token(&app).await?;
    let request = WarnChatUserRequest::new(
        params.broadcaster_id.as_str(),
        token.user_id.as_str(),
    );
    let body = WarnChatUserBody::new(params.user_id.as_str(), params.reason.as_str());
    helix()
        .req_post(request, body, &token)
        .await
        .map(|_| ())
        .map_err(|e| e.to_string())
}

// https://dev.twitch.tv/docs/api/reference/#manage-held-automod-messages
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ManageHeldAutomodMessageParams {
    pub msg_id: String,
}

#[tauri::command]
pub async fn approve_held_automod_message(
    app: tauri::AppHandle,
    params: ManageHeldAutomodMessageParams,
) -> Result<(), String> {
    manage_held_automod_message(&app, params.msg_id.as_str(), true).await
}

#[tauri::command]
pub async fn deny_held_automod_message(
    app: tauri::AppHandle,
    params: ManageHeldAutomodMessageParams,
) -> Result<(), String> {
    manage_held_automod_message(&app, params.msg_id.as_str(), false).await
}

async fn manage_held_automod_message(
    app: &tauri::AppHandle,
    msg_id: &str,
    allow: bool,
) -> Result<(), String> {
    let token = get_token(app).await?;
    let request = ManageHeldAutoModMessagesRequest::new();
    let body = ManageHeldAutoModMessagesBody::new(token.user_id.as_str(), msg_id, allow);
    helix()
        .req_post(request, body, &token)
        .await
        .map(|_| ())
        .map_err(|e| e.to_string())
}

File: src/commands/moderation.ts
import { invokeCommand, type InvokeOptions, type Paginated } from "./utils";

export type DeleteChatMessagesParams = {
  broadcasterId: string;
  messageId?: string | null;
};
export type DeleteChatMessagesResponse = void;

export function deleteChatMessages(
  params: DeleteChatMessagesParams,
  options?: InvokeOptions,
): Promise<DeleteChatMessagesResponse> {
  const successMessage = params.messageId ? "Message deleted" : "Chat cleared";
  return invokeCommand("delete_chat_messages", params, { successMessage, ...options });
}

export type BanUserParams = {
  broadcasterId: string;
  userId: string;
  duration?: number | null;
  reason?: string | null;
};
export type BanUserResponse = {
  broadcaster_id: string;
  moderator_id: string;
  user_id: string;
  created_at: string;
  end_time: string | null;
};

export function banUser(
  params: BanUserParams,
  options?: InvokeOptions,
): Promise<BanUserResponse> {
  const successMessage = params.duration ? "User timed out" : "User banned";
  return invokeCommand("ban_user", params, { successMessage, ...options });
}

export type UnbanUserParams = {
  broadcasterId: string;
  userId: string;
};
export type UnbanUserResponse = void;

export function unbanUser(
  params: UnbanUserParams,
  options?: InvokeOptions,
): Promise<UnbanUserResponse> {
  return invokeCommand("unban_user", params, { successMessage: "User unbanned", ...options });
}

export type GetBannedUsersParams = {
  broadcasterId: string;
  userId?: string;
  first?: number;
  after?: string;
};
export type GetBannedUsersResponse = Paginated<{
  user_id: string;
  user_login: string;
  user_name: string;
  expires_at: string | null;
  created_at: string;
  reason: string;
  moderator_id: string;
  moderator_login: string;
  moderator_name: string;
}>;

export function getBannedUsers(
  params: GetBannedUsersParams,
  options?: InvokeOptions,
): Promise<GetBannedUsersResponse> {
  return invokeCommand("get_banned_users", params, options);
}

export type GetAllBannedUsersParams = {
  broadcasterId: string;
  userId?: string;
};

export function getAllBannedUsers(
  params: GetAllBannedUsersParams,
  options?: InvokeOptions,
): Promise<GetBannedUsersResponse["data"]> {
  return invokeCommand("get_all_banned_users", params, options);
}

export type GetModeratorsParams = {
  broadcasterId: string;
  first?: number;
  after?: string;
};
export type GetModeratorsResponse = Paginated<{
  user_id: string;
  user_login: string;
  user_name: string;
}>;

export function getModerators(
  params: GetModeratorsParams,
  options?: InvokeOptions,
): Promise<GetModeratorsResponse> {
  return invokeCommand("get_moderators", params, options);
}

export type GetAllModeratorsParams = {
  broadcasterId: string;
};

export function getAllModerators(
  params: GetAllModeratorsParams,
  options?: InvokeOptions,
): Promise<GetModeratorsResponse["data"]> {
  return invokeCommand("get_all_moderators", params, options);
}

export type GetModeratedChannelsParams = {
  first?: number;
  after?: string;
};
export type GetModeratedChannelsResponse = Paginated<{
  broadcaster_id: string;
  broadcaster_login: string;
  broadcaster_name: string;
}>;

export function getModeratedChannels(
  params: GetModeratedChannelsParams = {},
  options?: InvokeOptions,
): Promise<GetModeratedChannelsResponse> {
  return invokeCommand("get_moderated_channels", params, options);
}

export function getAllModeratedChannels(
  options?: InvokeOptions,
): Promise<GetModeratedChannelsResponse["data"]> {
  return invokeCommand("get_all_moderated_channels", undefined, options);
}

export type AddChannelVipParams = {
  broadcasterId: string;
  userId: string;
};

export function addChannelVip(
  params: AddChannelVipParams,
  options?: InvokeOptions,
): Promise<void> {
  return invokeCommand("add_channel_vip", params, { successMessage: "VIP added", ...options });
}

export type RemoveChannelVipParams = {
  broadcasterId: string;
  userId: string;
};

export function removeChannelVip(
  params: RemoveChannelVipParams,
  options?: InvokeOptions,
): Promise<void> {
  return invokeCommand("remove_channel_vip", params, { successMessage: "VIP removed", ...options });
}

export type StartRaidParams = {
  fromBroadcasterId: string;
  toBroadcasterId: string;
};

export function startRaid(
  params: StartRaidParams,
  options?: InvokeOptions,
): Promise<void> {
  return invokeCommand("start_raid", params, { successMessage: "Raid started", ...options });
}

export type CancelRaidParams = {
  broadcasterId: string;
};

export function cancelRaid(
  params: CancelRaidParams,
  options?: InvokeOptions,
): Promise<void> {
  return invokeCommand("cancel_raid", params, { successMessage: "Raid cancelled", ...options });
}

export type WarnUserParams = {
  broadcasterId: string;
  userId: string;
  reason: string;
};

export function warnUser(
  params: WarnUserParams,
  options?: InvokeOptions,
): Promise<void> {
  return invokeCommand("warn_user", params, { successMessage: "User warned", ...options });
}

export type ManageHeldAutomodMessageParams = {
  msgId: string;
};

export function approveHeldAutomodMessage(
  params: ManageHeldAutomodMessageParams,
  options?: InvokeOptions,
): Promise<void> {
  return invokeCommand("approve_held_automod_message", params, {
    successMessage: "Message approved",
    ...options,
  });
}

export function denyHeldAutomodMessage(
  params: ManageHeldAutomodMessageParams,
  options?: InvokeOptions,
): Promise<void> {
  return invokeCommand("deny_held_automod_message", params, {
    successMessage: "Message denied",
    ...options,
  });
}
